File: tugrust/crates/tugtool/src/session_identity.rs
//! The calling session's identity — **the one place `$TUG_SESSION_ID` is
//! read**, and the one place a posted id becomes a session the machine will
//! answer for ([P01]).
//!
//! `TUG_SESSION_ID` is frozen in a process's environment at spawn, and the
//! Wheel rotates a card's session *on purpose* — above
//! `implement_compact_tokens`, on a threshold any long implement stage
//! crosses. So the id a short-lived `tugtool` process was born with routinely
//! names a segment that closed and was demoted two rotations ago. A verb that
//! uses it raw writes onto a corpse and reports success: a truthful sentence
//! about the wrong session, which is exactly how an arc came to be stranded
//! while every gesture said it had worked
//! (`notes/wheel-rotation-strands-the-arc.md`).
//!
//! The answer is not "expand it at each door" — that was tried, once per
//! caller, and each fix left the next caller exposed. The answer is that the
//! raw id is **unusable**: every verb calls [`resolve`] (or [`resolve_soft`]),
//! and `tugcore`'s `no_raw_session_id_reads` source scan refuses a second
//! reader of the variable anywhere in the workspace. The chokepoint is
//! structural, so the next verb cannot forget.
//!
//! The expansion itself belongs to the instance that owns the line —
//! `sessions.db` is per-instance — so this asks it, over the same
//! try-each-instance POST the binding verbs use, at `POST /api/session`
//! `{op: "resolve"}`.

use crate::dash::post_instance_api;

/// A resolved calling session: what was posted, what it resolved to, and
/// enough of the ledger's answer to write a sentence about either.
#[derive(Debug, Clone)]
pub(crate) struct Resolved {
    /// The id the process was born with (or `--session`'s argument).
    pub posted: String,
    /// The live segment of the posted id's line — the session every verb
    /// addresses. Equal to `posted` in the ordinary un-rotated case, and on
    /// the soft fallback below.
    pub session_id: String,
    /// The resolved segment's ledger `state`, or `None` when no instance
    /// answered.
    pub state: Option<String>,
    /// The line the posted id belongs to, when an instance placed it.
    pub line_id: Option<String>,
    /// The checkout this session works, as the ledger row spells it.
    ///
    /// `None` from an instance that predates the field — and from the soft
    /// fallback, where nobody answered at all. A caller uses it to tell "this
    /// session could have done that and did not" from "this session was never
    /// entitled to", and must treat `None` as "cannot tell".
    pub project_dir: Option<String>,
    /// Every segment id the line has worn — what a session-keyed *reader*
    /// expands over. Empty when nothing answered.
    pub segments: Vec<String>,
    /// Whether the resolution actually moved: the posted id named a segment
    /// that is no longer the seated one.
    pub rotated: bool,
    /// Whether an instance answered at all. `false` is the soft fallback:
    /// the posted id, unexpanded, because there was nobody to ask.
    pub resolved: bool,
}

impl Resolved {
    /// The posted id, unexpanded — the answer when no instance could be
    /// asked. Never reached by a verb that needs an instance anyway.
    fn unresolved(posted: String) -> Self {
        Resolved {
            session_id: posted.clone(),
            posted,
            state: None,
            line_id: None,
            project_dir: None,
            segments: Vec::new(),
            rotated: false,
            resolved: false,
        }
    }

    /// Every key a session-addressed row may sit under, most-current first:
    /// the live segment, then the rest of the line, then the posted id. A
    /// reader probes these in order so a row written under an earlier segment
    /// still surfaces after a rotation.
    pub(crate) fn keys(&self) -> Vec<String> {
        let mut keys = vec![self.session_id.clone()];
        for id in self.segments.iter().chain(std::iter::once(&self.posted)) {
            if !keys.contains(id) {
                keys.push(id.clone());
            }
        }
        keys
    }
}

/// The id this process was born with — **the only read of `TUG_SESSION_ID`
/// in the workspace outside the spawn-time exporters.**
///
/// Private on purpose: a caller that wants the id wants the *resolved* one,
/// and the two functions below are the only ways to reach it.
fn posted_session_id() -> Option<String> {
    std::env::var("TUG_SESSION_ID")
        .ok()
        .filter(|s| !s.is_empty())
}

/// Whether this process was born with a calling session at all.
///
/// The one question about the raw id that is not about its *value*, and so the
/// one a caller may ask without resolving. A verb whose claim is best-effort
/// asks it to tell "there is nobody to claim for" — a headless run, a fixture,
/// a plain terminal — apart from "somebody was asked and said no". The id
/// itself does not leave this module.
pub(crate) fn have_calling_session() -> bool {
    posted_session_id().is_some()
}

/// The calling session, resolved to its line's live segment, or the
/// actionable refusal naming what to do.
///
/// `subject` names what wanted the session, so the refusal says why it is
/// asking rather than reporting whatever the first caller happened to be.
/// `session` is `--session`'s argument, which overrides the environment and
/// is resolved on exactly the same terms — an id typed by hand goes stale the
/// same way one inherited from a spawn does.
pub(crate) fn resolve(subject: &str, session: Option<&str>) -> Result<Resolved, String> {
    let posted = session
        .map(str::to_string)
        .filter(|s| !s.is_empty())
        .or_else(posted_session_id)
        .ok_or_else(|| {
            format!(
                "no session — {subject} names the calling session, so run this from a Session \
                 card, pass --session <id>, or set TUG_SESSION_ID"
            )
        })?;
    ask_instance(&posted).map_err(|message| {
        // `unknown_session` is the walk's "not mine" answer; when *every*
        // instance says it, the id belongs to no machine that is running.
        if message == "unknown_session" {
            // The trailing token is a **contract**, not leakage. This refusal
            // has one legitimately transient cause — the ledger row a card's
            // spawn writes has not landed yet — and a caller that must sit
            // through that window needs to tell it from a permanent refusal.
            // Before this module the walk's raw `unknown_session` reached
            // stderr and callers branched on it; replacing it with prose alone
            // broke the app-test fixture's retry loop
            // (`tests/app-test/dash-fixture.ts`'s `bindDash`) silently, and the
            // breakage surfaced three workstreams downstream. So the token
            // stays, and the CLI test below fails in `cargo nextest` the next
            // time somebody rewrites the sentence.
            format!(
                "no running Tug instance knows session {posted} — {subject} needs the card that \
                 session works, and nothing here is holding it (unknown_session)"
            )
        } else {
            format!("{subject} could not resolve session {posted}: {message}")
        }
    })
}

/// [`resolve`], but a machine that cannot be asked is not an error.
///
/// For the verbs that still mean something with no instance running — a
/// `draft` under `TUG_CHANGES_DB`, a `changes` read straight off the ledger.
/// They get the posted id back with `resolved: false`, which is the same
/// answer they had before this module existed, reached deliberately instead
/// of by omission.
pub(crate) fn resolve_soft(session: Option<&str>) -> Option<Resolved> {
    let posted = session
        .map(str::to_string)
        .filter(|s| !s.is_empty())
        .or_else(posted_session_id)?;
    Some(ask_instance(&posted).unwrap_or_else(|_| Resolved::unresolved(posted)))
}

/// Ask the owning instance a session-addressed question **whose op resolves at
/// its own door**, with the posted id filled in from this process's identity.
///
/// The chokepoint's one narrow opening, and what makes it safe: no caller ever
/// receives the raw id, and every op reached this way expands it server-side
/// exactly as `/api/dash` does. What it buys is a single round trip instead of
/// a resolve followed by the real question — which matters on the hot path,
/// where the PreToolUse gate asks on every write-shaped tool call a course
/// stage makes.
///
/// The outer `None` is "this process has no calling session at all" — a plain
/// terminal, a fixture, a foreign project — which is not a failure and must
/// never read as one. The inner `Err` is an instance that answered and refused,
/// including `unknown op` from one older than the op; every caller of this is
/// advisory ([P01]'s skew rule, Part IV item 4) and must degrade rather than
/// refuse on it.
pub(crate) fn ask_about_calling_session(
    op: &str,
    subject: &str,
    fields: serde_json::Value,
) -> Option<Result<serde_json::Value, String>> {
    let posted = posted_session_id()?;
    let mut body = serde_json::json!({ "op": op, "tug_session_id": posted });
    if let (Some(target), Some(extra)) = (body.as_object_mut(), fields.as_object()) {
        for (key, value) in extra {
            target.insert(key.clone(), value.clone());
        }
    }
    Some(post_instance_api("/api/session", subject, body))
}
/// Ask the instance that owns the line. Walks every live instance, exactly as
/// a binding write does: `sessions.db` is per-instance, so the first machine
/// to answer is not always the right one.
fn ask_instance(posted: &str) -> Result<Resolved, String> {
    let response = post_instance_api(
        "/api/session",
        "resolving the calling session",
        serde_json::json!({ "op": "resolve", "tug_session_id": posted }),
    );
    let response = match response {
        Ok(response) => response,
        // An instance older than the chokepoint answers `unknown op
        // 'resolve'`. Refusing there would break every session-addressed verb
        // the moment a new `tugtool` met a tugcast that had not restarted —
        // and it is not necessary: `/api/dash` resolves at its own door, so a
        // bind still lands on the live segment. What is lost is this side's
        // ability to *name* the resolution, which is a receipt, not a write.
        Err(message) if message.contains("unknown op") => {
            return Ok(Resolved::unresolved(posted.to_string()));
        }
        Err(message) => return Err(message),
    };
    let str_field = |key: &str| {
        response
            .get(key)
            .and_then(|v| v.as_str())
            .map(str::to_string)
    };
    let session_id = str_field("session_id").unwrap_or_else(|| posted.to_string());
    Ok(Resolved {
        rotated: session_id != posted,
        posted: posted.to_string(),
        session_id,
        state: str_field("state"),
        line_id: str_field("line_id"),
        project_dir: str_field("project_dir"),
        segments: response
            .get("segments")
            .and_then(|v| v.as_array())
            .map(|a| {
                a.iter()
                    .filter_map(|v| v.as_str().map(str::to_string))
                    .collect()
            })
            .unwrap_or_default(),
        resolved: true,
    })
}

File: tugdeck/src/lib/session-identity.ts
/**
 * session-identity.ts — the one place Tug decides what a session is called.
 *
 * Tug used to name a session five different ways: a chip rule, a chooser-row
 * rule, a card-row hash sniff, a title-bar composer, and the Rust feed's own
 * precedence. No two agreed. This module is [D123] ("a pane's name is one
 * string produced in one place") applied to the session: one resolver produces
 * one structured {@link SessionIdentity} record, and every surface renders that
 * record at a declared density tier.
 *
 * Three entry points, and which one you want is not a matter of taste:
 *
 *  - {@link useSessionIdentity} — **the React entry point.** It subscribes to
 *    each backing store *by id* and then derives. Every component uses this.
 *  - {@link resolveSessionIdentity} — an imperative SNAPSHOT with no
 *    subscription. Correct only for non-React callers: clipboard writes,
 *    export filenames, command routes, list projections that already carry
 *    the stores' version tokens as inputs.
 *  - {@link sessionIdentityLine} / {@link sessionCitation} — pure formatters
 *    over a resolved record.
 *
 * **Calling the bare resolver from a render is the bug this module exists to
 * prevent.** It reads three stores imperatively; from a render body it hands
 * back a snapshot with no subscription, so a `/rename` or a synopsis write
 * updates the stores while the surface repaints never. That failure passes
 * every unit test and shows up only as a stale name in the running app, which
 * is why "no component calls `resolveSessionIdentity`" is a grep gate rather
 * than a convention.
 *
 * **Liveness is not identity.** The record carries no phase field. Phase lives
 * on a per-*card* `codeSessionStore`, and folding it in would wake every
 * identity surface on every transcript event. `useSessionPhase` is the
 * session-keyed door onto it, returning a short key rather than the snapshot,
 * and the phase dot stays a leaf subscription composed *into* the chip, row, and
 * masthead tiers as a child; identity and liveness are two subscriptions with
 * two keys that meet in the component.
 *
 * **The branch is not identity either.** It is telemetry — it rides the record
 * for the masthead's placard and never appears in a rendered name. The old
 * `(<branch>)` title suffix is retired.
 *
 * **Resolvability is a field, not an inference from `tag`.** A citation may name
 * a session this ledger has never held, and the honest rendering still shows the
 * callsign that citation recorded ([P13]) — so `tag` alone cannot answer "can
 * this be found", and {@link SessionIdentity.resolved} does. Whether the ledger
 * holds a cited session is asked over the wire by
 * `session-citation-store.ts`; this module only records the answer.
 *
 * Laws: [L02] external state enters React through `useSyncExternalStore` only.
 *
 * @module lib/session-identity
 */

import { useCallback, useSyncExternalStore } from "react";

import {
  cardSessionBindingStore,
  type CardSessionBinding,
} from "@/lib/card-session-binding-store";
import { sessionNameStore } from "@/lib/session-name-store";
import { sessionSynopsisStore } from "@/lib/session-synopsis-store";
import { sessionTagStore } from "@/lib/session-tag-store";
import {
  identityKeyForSession,
  sessionLineStore,
} from "@/lib/session-line-store";
import type { SessionRow } from "@/protocol";

/** Chars of a session UUID that make its short id. Computed here and nowhere else. */
export const SESSION_SHORT_ID_LENGTH = 8;

/**
 * One session's identity, as every surface in the app reads it.
 *
 * Deliberately absent: phase/liveness (a per-card subscription, composed in as
 * a child) and anything a caller would have to re-derive.
 */
export interface SessionIdentity {
  /** Project leaf name (basename of projectDir); `""` when the project is unknown. */
  project: string;
  /** Workspace branch, or null — telemetry only, never rendered in identity. */
  branch: string | null;
  /** The callsign; null only for a legacy tagless row. Stable for the life
   *  of the line of work — a rewind-fork inherits it by transfer. */
  tag: string | null;
  /**
   * The user's own name for this session — `/rename`, never an auto title.
   *
   * Independent of {@link description}, and never merged with it: the name is
   * the title a surface leads with ({@link sessionTitleParts}) and the
   * description is the line beneath it. A single merged field could serve
   * neither.
   */
  customName: string | null;
  /** The agent's rolling description. Independent of {@link customName}. */
  description: string | null;
  /** Ledger state, when the caller knows it. */
  state: SessionRow["state"] | null;
  /** Full tug session id (plumbing: tooltips, copy affordance). */
  id: string;
  /** First {@link SESSION_SHORT_ID_LENGTH} chars of `id` — THE short id. */
  shortId: string;
  /**
   * Whether this identity is the LEDGER's word about a session it actually
   * holds, as against a reference reconstructed from what a commit recorded.
   *
   * This is the resolvable/unresolvable distinction, and it has to live on the
   * record because `tag` cannot carry it: {@link SessionIdentityContext.recordedTag}
   * fills the callsign for a citation the ledger has never heard of, precisely
   * so an unresolvable reference can still say what it named ([P13]). A reader
   * of `tag` alone therefore cannot tell the two apart, and one that tried
   * would call every foreign commit's session findable.
   *
   * `true` when some ledger-fed source named this session: a callsign in the
   * tag store, a card binding supplying its project, or a `resolve_sessions`
   * answer. Note what it is NOT: liveness. A closed session is resolved; a
   * commit from another machine is not. Sessions are never dead, only
   * unfindable ([P13], [P15]).
   */
  resolved: boolean;
}

/**
 * The facts about a session that do NOT live in the three by-id identity
 * stores, supplied by a caller that already holds them.
 *
 * The picker and the Cards card hold a whole `SessionRow` and pass it through
 * {@link sessionIdentityContextFrom}; the Session card and the masthead pass
 * their project dir and (for the placard) the branch. Anything omitted
 * resolves to its honest empty value rather than being invented — except
 * `projectDir`, which falls back to the session's card binding when one
 * exists.
 */
export interface SessionIdentityContext {
  projectDir?: string | null;
  branch?: string | null;
  state?: SessionRow["state"] | null;
  /**
   * A callsign recorded **elsewhere**, at write time — the tag inside a commit's
   * `Tug-Session:` trailer. Used only when this ledger has none of its own, so
   * the ledger stays authoritative and a rerolled callsign is never overruled by
   * an older commit's copy of it.
   *
   * It exists because a citation can be unresolvable and still informative: a
   * commit made on another machine names a session this ledger has no record of,
   * and the honest rendering is the callsign that commit actually recorded,
   * slashed and inert ([P13]), rather than a bare hash the commit never said.
   *
   * It fills `tag` and deliberately does **not** touch
   * {@link SessionIdentity.resolved} — that separation is the whole point. A
   * recorded callsign is what the reference *said*, never evidence that this
   * ledger can find what it named.
   */
  recordedTag?: string | null;
  /**
   * A project leaf-name recorded **elsewhere** — the `<project>` head of a
   * session atom's `<project>/<callsign>` value. The same contract as
   * {@link recordedTag}, for the other half of the reference: it fills
   * {@link SessionIdentity.project} only when no ledger-fed source supplies a
   * project dir, and it never makes the identity
   * {@link SessionIdentity.resolved} — what a reference recorded is what it
   * said, never evidence that this ledger can find what it named.
   */
  recordedProject?: string | null;
  /**
   * The ledger's own answer that it holds this session, from a
   * `resolve_sessions` round trip (`session-citation-store.ts`).
   *
   * A citation surface has no card binding and may sit outside any listing, so
   * without this the two local signals would call every foreign commit's
   * session unfindable — and every locally-known one findable only by accident
   * of what had been listed. Omitted by callers who hold no such answer, which
   * leaves resolvability to those local signals exactly as before.
   */
  ledgerKnown?: boolean;
}

/**
 * The trailing path component (basename) of a directory: the project's
 * leaf-name identity. Trailing slashes are ignored; an empty or slash-only
 * path yields the trimmed input.
 */
export function projectLeafName(dir: string): string {
  const trimmed = dir.replace(/\/+$/, "");
  const leaf = trimmed.split("/").pop() ?? "";
  return leaf.length > 0 ? leaf : trimmed;
}

/** The short id: the first {@link SESSION_SHORT_ID_LENGTH} chars of the UUID. */
export function shortSessionId(sessionId: string): string {
  return sessionId.slice(0, SESSION_SHORT_ID_LENGTH);
}

/**
 * Compose an identity record from explicit facts — the pure half, with no
 * store reads. This is what the unit tests exercise; {@link resolveSessionIdentity}
 * is this function with the three stores read for it.
 */
export function composeSessionIdentity(input: {
  sessionId: string;
  name: string | null;
  synopsis: string | null;
  /** The callsign THIS LEDGER holds for the session, or null. */
  tag: string | null;
  /**
   * The callsign a commit recorded, used only when the ledger has none. It
   * fills `tag` without making the identity {@link SessionIdentity.resolved} —
   * see the field's own note.
   */
  recordedTag?: string | null;
  /**
   * A project leaf-name the reference recorded. Fills `project` only when
   * `projectDir` supplies none, and never affects `resolved` — see the
   * context field's own note.
   */
  recordedProject?: string | null;
  projectDir?: string | null;
  branch?: string | null;
  state?: SessionRow["state"] | null;
  /** The ledger's explicit answer that it holds this session. */
  ledgerKnown?: boolean;
}): SessionIdentity {
  const ledgerTag = input.tag?.trim() || null;
  const recordedTag = input.recordedTag?.trim() || null;
  const tag = ledgerTag ?? recordedTag;
  const name = input.name?.trim() || null;
  const synopsis = input.synopsis?.trim() || null;
  const projectDir = input.projectDir?.trim() ?? "";
  const recordedProject = input.recordedProject?.trim() || null;
  const branch = input.branch?.trim() || null;
  return {
    project:
      projectDir.length > 0
        ? projectLeafName(projectDir)
        : (recordedProject ?? ""),
    branch,
    tag,
    // Resolvability rests only on ledger-fed facts. `recordedTag` is
    // conspicuously absent from this expression: a commit's own copy of a
    // callsign tells the reader what was named, never that it can be found.
    resolved:
      input.ledgerKnown === true || ledgerTag !== null || projectDir.length > 0,
    // Two fields, never a fallback between them: they occupy different lines on
    // every surface. The name store holds only user-set names, so an auto title
    // can never arrive here as a `customName`.
    customName: name,
    description: synopsis,
    state: input.state ?? null,
    id: input.sessionId,
    shortId: shortSessionId(input.sessionId),
  };
}

/** The context a `SessionRow` supplies — the picker's and the Cards card's path in. */
export function sessionIdentityContextFrom(
  row: SessionRow,
): SessionIdentityContext {
  return {
    projectDir: row.project_dir,
    state: row.state,
  };
}

/** The project dir of whichever card is bound to `sessionId`, or null. */
function boundProjectDirFor(sessionId: string): string | null {
  const lineId = sessionLineStore.lineOf(sessionId);
  for (const binding of cardSessionBindingStore.getSnapshot().values()) {
    if (binding.tugSessionId === sessionId) return binding.projectDir;
    if (lineId !== null && binding.lineId === lineId) return binding.projectDir;
  }
  return null;
}

/**
 * Resolve a session's identity **as a snapshot, with no subscription**.
 *
 * Non-React callers only — clipboard writes, export filenames, command routes,
 * and list projections that already carry the identity stores' version tokens
 * as recompute inputs. Every component uses {@link useSessionIdentity} instead;
 * calling this from a render is the stale-render bug this module's header
 * describes, and a grep gate exists to keep it out of `components/`.
 */
export function resolveSessionIdentity(
  sessionId: string,
  context?: SessionIdentityContext,
): SessionIdentity {
  // The name, the callsign and the synopsis belong to the **line** ([P12]) —
  // the segment id is only the address the caller happens to hold.
  const lineId = identityKeyForSession(sessionId);
  return composeSessionIdentity({
    sessionId,
    name: sessionNameStore.getName(lineId),
    synopsis: sessionSynopsisStore.getSynopsis(lineId),
    tag: sessionTagStore.getTag(lineId),
    recordedTag: context?.recordedTag ?? null,
    recordedProject: context?.recordedProject ?? null,
    projectDir: context?.projectDir ?? boundProjectDirFor(sessionId),
    branch: context?.branch ?? null,
    state: context?.state ?? null,
    ledgerKnown: context?.ledgerKnown ?? false,
  });
}

/**
 * The React entry point: subscribe to each backing store **by id**, then
 * resolve. Returns null for a null `sessionId` so a caller with no session
 * needs no branch of its own.
 *
 * By-id getters, deliberately: the stores also publish whole-store version
 * tokens, but those exist for consumers deriving across many sessions at once
 * (the Cards card's filter projection). Reading a version token here would wake every
 * identity surface in the app on any session's rename.
 *
 * Overloaded on the argument, so a caller that already holds a session id gets
 * a record rather than a maybe-record and writes no branch for a state its own
 * types forbid. The nullable arm exists for callers whose binding may not have
 * resolved yet.
 */
export function useSessionIdentity(
  sessionId: string,
  context?: SessionIdentityContext,
): SessionIdentity;
export function useSessionIdentity(
  sessionId: string | null,
  context?: SessionIdentityContext,
): SessionIdentity | null;
export function useSessionIdentity(
  sessionId: string | null,
  context?: SessionIdentityContext,
): SessionIdentity | null {
  // The line the segment belongs to, subscribed like the rest: a rotation
  // re-keys this card's identity, and a hook that read the mapping
  // unsubscribed would keep resolving under the id the card has left.
  const lineId = useSyncExternalStore(
    sessionLineStore.subscribe,
    useCallback(
      () => (sessionId === null ? null : identityKeyForSession(sessionId)),
      [sessionId],
    ),
  );
  const name = useSyncExternalStore(
    sessionNameStore.subscribe,
    useCallback(
      () => (lineId === null ? null : sessionNameStore.getName(lineId)),
      [lineId],
    ),
  );
  const tag = useSyncExternalStore(
    sessionTagStore.subscribe,
    useCallback(
      () => (lineId === null ? null : sessionTagStore.getTag(lineId)),
      [lineId],
    ),
  );
  const synopsis = useSyncExternalStore(
    sessionSynopsisStore.subscribe,
    useCallback(
      () =>
        lineId === null
          ? null
          : sessionSynopsisStore.getSynopsis(lineId),
      [lineId],
    ),
  );
  // The project dir a caller did not supply. The snapshot is the dir STRING,
  // not the binding, so React bails out on every binding change that does not
  // move this session's project — which is all of them but a rebind.
  const boundProjectDir = useSyncExternalStore(
    cardSessionBindingStore.subscribe,
    useCallback(
      () => (sessionId === null ? null : boundProjectDirFor(sessionId)),
      [sessionId],
    ),
  );
  if (sessionId === null) return null;
  return composeSessionIdentity({
    sessionId,
    name,
    synopsis,
    tag,
    recordedTag: context?.recordedTag ?? null,
    recordedProject: context?.recordedProject ?? null,
    projectDir: context?.projectDir ?? boundProjectDir,
    branch: context?.branch ?? null,
    state: context?.state ?? null,
    ledgerKnown: context?.ledgerKnown ?? false,
  });
}

/**
 * The two runs a surface renders as a session's title.
 *
 * **A custom name means no callsign, full stop.** A user-supplied name is the
 * user saying what the session is called, and the name REPLACES the callsign
 * rather than leading it: showing both harmed legibility and made the title
 * too long for the interface beside it. The callsign stays the permanent
 * citable handle — the tooltip, the citation, and every copy path carry it
 * whole — it just never rides the title.
 *
 * There is no collision exception, because a collision can no longer occur:
 * a user-set name is unique across lines at the write, and a name another line
 * already wears is TAKEN from it by the newer `/rename` — the previous holder
 * falls back to its callsign ([D167]). Uniqueness at the write is what retired
 * the exception; the callsign the reader would have needed to disambiguate
 * with is a callsign there is nothing to disambiguate.
 *
 * With no name at all, the callsign IS the title and there is no second run.
 * The `callsign` run therefore appears only for an unnamed session, which is
 * also why it carries the `project/` prefix — `tugtool/frothy-nurse` — the
 * project a session works against being how a reader places it in a list of
 * sessions from many projects. The composed run is exactly
 * {@link sessionIdentityLine}, so the title and the Line channel cannot spell
 * the same session two ways.
 *
 * Two runs rather than one joined string, because they are sized separately.
 * The `":"` separator belongs to the callsign run so it disappears with it.
 */
export function sessionTitleParts(identity: SessionIdentity): {
  name: string;
  callsign: string | null;
} {
  return {
    name: identity.customName ?? sessionIdentityLine(identity),
    callsign: null,
  };
}

/**
 * The same title as ONE string, for a surface that cannot size two runs.
 *
 * {@link sessionTitleParts} exists because a measured surface wants to decide
 * which run gives way. A flat row, a baked chip, and an aria label have one
 * string and one truncation, and they still owe the reader the same rule — so
 * the join lives here rather than being spelled at each call site, where the
 * three would eventually disagree about how the two runs meet.
 *
 * The separator is the one the two-run form renders, and it belongs to the
 * callsign exactly as it does there: no callsign, no separator. Under the
 * unconditional rule there is never a second run, so this is the name for a
 * named session and the identity line for an unnamed one.
 */
export function sessionDisplayTitle(identity: SessionIdentity): string {
  const { name, callsign } = sessionTitleParts(identity);
  return callsign === null ? name : `${name}:${callsign}`;
}

/**
 * The Line-tier string: `<project>/<callsign>`.
 *
 * No branch suffix and no name arm — which is what makes this string
 * **constant for the life of a session binding**, and that is load-bearing
 * rather than incidental. The pane-title channel carries it for reader
 * compatibility (the tab strip, the Window menu, the slot-stack picker); it is
 * not a notification path, and identity changes reach surfaces through
 * {@link useSessionIdentity}. A legacy tagless session degrades to its short
 * id — never the full UUID.
 */
export function sessionIdentityLine(identity: SessionIdentity): string {
  const label = identity.tag ?? identity.shortId;
  return identity.project.length > 0
    ? `${identity.project}/${label}`
    : label;
}

/**
 * The citation — the only sanctioned flat-text form of a session reference,
 * and the only place monospace appears: `<tag> (<shortId>)`, project-prefixed
 * when the surrounding context does not supply the project.
 *
 * A legacy tagless session degrades to the bare short id with no parentheses:
 * the hash is already the callsign there, and printing it twice is noise.
 */
export function sessionCitation(
  identity: SessionIdentity,
  opts?: { project?: boolean },
): string {
  const withProject = opts?.project === true && identity.project.length > 0;
  if (identity.tag === null) {
    return withProject
      ? `${identity.project}/${identity.shortId}`
      : identity.shortId;
  }
  const head = withProject
    ? `${identity.project}/${identity.tag}`
    : identity.tag;
  return `${head} (${identity.shortId})`;
}

/** The parenthesized token of a citation: `<tag> (<token>)`. */
const CITATION_TOKEN = /\(([0-9a-fA-F-]+)\)\s*$/;

/** A full session UUID — the legacy trailer's token, and `Tug-Session-Id`'s. */
const FULL_UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

/** A short id: exactly {@link SESSION_SHORT_ID_LENGTH} hex chars. */
const SHORT_ID = /^[0-9a-f]{8}$/i;

/**
 * What a commit's session trailers name (Spec S03) — the id to resolve against,
 * and the callsign the commit itself recorded.
 */
export interface CitedSession {
  /**
   * The id the commit named, exactly as it named it: a full uuid when it
   * carries one (or a legacy trailer spelled one), else the citation's 8-char
   * short id.
   *
   * A short id is **not** expanded here. Expansion needs a view of every
   * session and the ability to notice that two share a prefix, and only the
   * ledger has both — `session-citation-store.ts` asks it, and hands back the
   * full id. A client-side prefix scan over whatever happened to be cached
   * would answer differently depending on which listings had run.
   */
  citedId: string;
  /** The callsign the trailer recorded, or null for a legacy tagless commit. */
  tag: string | null;
}

/**
 * Resolve a commit's session trailers to the session they name.
 *
 * Precedence per Spec S03, and the order is the point:
 *
 *  1. **`Tug-Session-Id`** — the machine field, a full uuid, an exact join.
 *     Every commit written after the trailer pair shipped carries it.
 *  2. **The citation's parenthesized token.** A 36-char uuid means a *legacy*
 *     one-line trailer (`<display> (<full-uuid>)`), which therefore resolves
 *     exactly too — legacy commits live in history forever and must not degrade
 *     to unresolvable. An 8-hex token is a short id, which the ledger expands.
 *  3. Neither parses → `null`, and the caller renders nothing at all.
 *
 * The callsign is read off the citation's head in both forms. For a legacy
 * trailer that head is a *display name*, not a tag — which is why it is only
 * ever a fallback ({@link SessionIdentityContext.recordedTag}) and never
 * overrules what this ledger says the session is called.
 *
 * Pure grammar: no store reads, so a caller may memoize it on the trailer text
 * alone. Turning the id into a *session* is the citation store's job.
 */
export function resolveCitedSession(
  citation: string | null | undefined,
  sessionId?: string | null,
): CitedSession | null {
  const raw = citation?.trim() ?? "";
  const machineId = sessionId?.trim() ?? "";
  const match = CITATION_TOKEN.exec(raw);
  // The tagless citation form is the bare short id with no parentheses, so a
  // trailer that is nothing but a token IS the token, and carries no callsign.
  const parenthesized = match !== null;
  const head = parenthesized ? raw.slice(0, match.index).trim() : "";
  const token = parenthesized ? match[1] : raw;
  const tag = head.length > 0 ? head : null;

  if (FULL_UUID.test(machineId)) return { citedId: machineId, tag };
  if (FULL_UUID.test(token)) return { citedId: token, tag };
  if (SHORT_ID.test(token)) return { citedId: token.toLowerCase(), tag };
  return null;
}

/**
 * The Line string for a session, resolved and formatted in one call — the
 * non-React shorthand for a list projection or a command route. Components use
 * {@link useSessionIdentity} + {@link sessionIdentityLine} instead.
 */
export function sessionIdentityLineFor(
  sessionId: string,
  context?: SessionIdentityContext,
): string {
  return sessionIdentityLine(resolveSessionIdentity(sessionId, context));
}

/** The Line string for a card's session binding — the Cards card's own projection. */
export function sessionIdentityLineForBinding(
  binding: CardSessionBinding,
): string {
  return sessionIdentityLineFor(binding.tugSessionId, {
    projectDir: binding.projectDir,
  });
}

/**
 * The display title for a card's session binding — the flat-row half of the
 * Cards card's own projection, beside {@link sessionIdentityLineForBinding}.
 *
 * The Cards card needs both and for different jobs: this is what a row SHOWS, the
 * Line is what a filter still MATCHES, so typing a callsign finds a session
 * whose row no longer prints one.
 */
export function sessionDisplayTitleForBinding(
  binding: CardSessionBinding,
): string {
  return sessionDisplayTitle(
    resolveSessionIdentity(binding.tugSessionId, {
      projectDir: binding.projectDir,
    }),
  );
}
